# Project 6: Analyzing Stock Sentiment from Twits
## Instructions
Each problem consists of a function to implement and instructions on how to implement the function.  The parts of the function that need to be implemented are marked with a `# TODO` comment.

## Packages
When you implement the functions, you'll only need to you use the packages you've used in the classroom, like [Pandas](https://pandas.pydata.org/) and [Numpy](http://www.numpy.org/). These packages will be imported for you. We recommend you don't add any import statements, otherwise the grader might not be able to run your code.

### Load Packages

In [1]:
import json
import nltk
import os
import random
import re
import torch

from torch import nn, optim
import torch.nn.functional as F

## Introduction
When deciding the value of a company, it's important to follow the news. For example, a product recall or natural disaster in a company's product chain. You want to be able to turn this information into a signal. Currently, the best tool for the job is a Neural Network. 

For this project, you'll use posts from the social media site [StockTwits](https://en.wikipedia.org/wiki/StockTwits). The community on StockTwits is full of investors, traders, and entrepreneurs. Each message posted is called a Twit. This is similar to Twitter's version of a post, called a Tweet. You'll build a model around these twits that generate a sentiment score.

We've collected a bunch of twits, then hand labeled the sentiment of each. To capture the degree of sentiment, we'll use a five-point scale: very negative, negative, neutral, positive, very positive. Each twit is labeled -2 to 2 in steps of 1, from very negative to very positive respectively. You'll build a sentiment analysis model that will learn to assign sentiment to twits on its own, using this labeled data.

The first thing we should to do, is load the data.

## Import Twits 
### Load Twits Data 
This JSON file contains a list of objects for each twit in the `'data'` field:

```
{'data':
  {'message_body': 'Neutral twit body text here',
   'sentiment': 0},
  {'message_body': 'Happy twit body text here',
   'sentiment': 1},
   ...
}
```

The fields represent the following:

* `'message_body'`: The text of the twit.
* `'sentiment'`: Sentiment score for the twit, ranges from -2 to 2 in steps of 1, with 0 being neutral.


To see what the data look like by printing the first 10 twits from the list. 

In [2]:
with open(os.path.join('..', '..', 'data', 'project_6_stocktwits', 'twits.json'), 'r') as f:
    twits = json.load(f)

print(twits['data'][:10])

[{'message_body': '$FITB great buy at 26.00...ill wait', 'sentiment': 2, 'timestamp': '2018-07-01T00:00:09Z'}, {'message_body': '@StockTwits $MSFT', 'sentiment': 1, 'timestamp': '2018-07-01T00:00:42Z'}, {'message_body': '#STAAnalystAlert for $TDG : Jefferies Maintains with a rating of Hold setting target price at USD 350.00. Our own verdict is Buy  http://www.stocktargetadvisor.com/toprating', 'sentiment': 2, 'timestamp': '2018-07-01T00:01:24Z'}, {'message_body': '$AMD I heard there’s a guy who knows someone who thinks somebody knows something - on StockTwits.', 'sentiment': 1, 'timestamp': '2018-07-01T00:01:47Z'}, {'message_body': '$AMD reveal yourself!', 'sentiment': 0, 'timestamp': '2018-07-01T00:02:13Z'}, {'message_body': '$AAPL Why the drop? I warren Buffet taking out his position?', 'sentiment': 1, 'timestamp': '2018-07-01T00:03:10Z'}, {'message_body': '$BA bears have 1 reason on 06-29 to pay more attention https://dividendbot.com?s=BA', 'sentiment': -2, 'timestamp': '2018-07-01T

### Length of Data
Now let's look at the number of twits in dataset. Print the number of twits below.

In [3]:
"""print out the number of twits"""

# TODO Implement 
print('[ info] Number of twits', len(twits['data']))


[ info] Number of twits 1548010


### Split Message Body and Sentiment Score

In [4]:
messages = [twit['message_body'] for twit in twits['data']]
# Since the sentiment scores are discrete, we'll scale the sentiments to 0 to 4 for use in our network
sentiments = [twit['sentiment'] + 2 for twit in twits['data']]

## Preprocessing the Data
With our data in hand we need to preprocess our text. These twits are collected by filtering on ticker symbols where these are denoted with a leader $ symbol in the twit itself. For example,

`{'message_body': 'RT @google Our annual look at the year in Google blogging (and beyond) http://t.co/sptHOAh8 $GOOG',
 'sentiment': 0}`

The ticker symbols don't provide information on the sentiment, and they are in every twit, so we should remove them. This twit also has the `@google` username, again not providing sentiment information, so we should also remove it. We also see a URL `http://t.co/sptHOAh8`. Let's remove these too.

The easiest way to remove specific words or phrases is with regex using the `re` module. You can sub out specific patterns with a space:

```python
re.sub(pattern, ' ', text)
```
This will substitute a space with anywhere the pattern matches in the text. Later when we tokenize the text, we'll split appropriately on those spaces.

### Pre-Processing

In [5]:
nltk.download('wordnet')
nltk.download('punkt')

def preprocess(message):
    """
    This function takes a string as input, then performs these operations: 
        - lowercase
        - remove URLs
        - remove ticker symbols 
        - removes punctuation
        - tokenize by splitting the string on whitespace 
        - removes any single character tokens
    
    Parameters
    ----------
        message : The text message to be preprocessed.
        
    Returns
    -------
        tokens: The preprocessed text into tokens.
    """ 
    #TODO: Implement 
    
    # Lowercase the twit message
    text = message.lower()
    
    # Replace URLs with a space in the message
    text = re.sub(r'http[s]?://(?:[a-zA-Z]|[0-9]|[$-_@.&+]|[!*\(\),]|(?:%[0-9a-fA-F][0-9a-fA-F]))+',' ', text)
    
    # Replace ticker symbols with a space. The ticker symbols are any stock symbol that starts with $.
    text = re.sub(r'[$][A-Za-z][\S]*', ' ', text)
    
    # Replace StockTwits usernames with a space. The usernames are any word that starts with @.
    text = re.sub(r'[@][A-Za-z][\S]*', ' ', text)

    # Replace everything not a letter with a space
    text = re.sub(r'[^A-Za-z]', ' ', text)
    
    # Tokenize by splitting the string on whitespace into a list of words
    #tokens = [word for word in text.split(' ') if len(word) > 0]
    tokens = nltk.word_tokenize(text)
    
    # Lemmatize words using the WordNetLemmatizer. You can ignore any word that is not longer than one character.
    wnl = nltk.stem.WordNetLemmatizer()
    tokens = [wnl.lemmatize(word) for word in tokens if len(word) > 1]
    
    return tokens

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Unzipping corpora/wordnet.zip.
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


### Preprocess All the Twits 
Now we can preprocess each of the twits in our dataset. Apply the function `preprocess` to all the twit messages.

In [6]:
# TODO Implement
tokenized = [preprocess(message) for message in messages]

In [7]:
tokenized[:5]

[['great', 'buy', 'at', 'ill', 'wait'],
 [],
 ['staanalystalert',
  'for',
  'jefferies',
  'maintains',
  'with',
  'rating',
  'of',
  'hold',
  'setting',
  'target',
  'price',
  'at',
  'usd',
  'our',
  'own',
  'verdict',
  'is',
  'buy'],
 ['heard',
  'there',
  'guy',
  'who',
  'know',
  'someone',
  'who',
  'think',
  'somebody',
  'know',
  'something',
  'on',
  'stocktwits'],
 ['reveal', 'yourself']]

### Bag of Words
Now with all of our messages tokenized, we want to create a vocabulary and count up how often each word appears in our entire corpus. Use the [`Counter`](https://docs.python.org/3.1/library/collections.html#collections.Counter) function to count up all the tokens.

In [8]:
from collections import Counter


"""
Create a vocabulary by using Bag of words
"""

# TODO: Implement 
vocab = [token for tokens in tokenized for token in tokens]
bow = Counter(vocab)

### Frequency of Words Appearing in Message
With our vocabulary, now we'll remove some of the most common words such as 'the', 'and', 'it', etc. These words don't contribute to identifying sentiment and are really common, resulting in a lot of noise in our input. If we can filter these out, then our network should have an easier time learning.

We also want to remove really rare words that show up in a only a few twits. Here you'll want to divide the count of each word by the number of messages. Then remove words that only appear in some small fraction of the messages.

In [9]:
import operator
"""
Set the following variables:
    freqs
    low_cutoff
    high_cutoff
    K_most_common
"""

# TODO Implement 

# Dictionart that contains the Frequency of words appearing in messages.
# The key is the token and the value is the frequency of that word in the corpus.
freqs = {word:bow[word]/len(bow) for id, word in enumerate(bow)}

# Float that is the frequency cutoff. Drop words with a frequency that is lower or equal to this number.
low_cutoff = 0.0000000001

# Integer that is the cut off for most common words. Drop words that are the `high_cutoff` most common words.
high_cutoff = 10

# The k most common words in the corpus. Use `high_cutoff` as the k.
K_most_common = [x[0] for x in sorted(freqs.items(), key=operator.itemgetter(1))[-high_cutoff:]]

filtered_words = [word for word in freqs if (freqs[word] > low_cutoff and word not in K_most_common)]
print(K_most_common)
len(filtered_words) 

['it', 'this', 'in', 'and', 'of', 'on', 'for', 'is', 'to', 'the']


99354

### Updating Vocabulary by Removing Filtered Words
Let's creat three variables that will help with our vocabulary.

In [10]:
"""
Set the following variables:
    vocab
    id2vocab
    filtered
"""

#TODO Implement
# A dictionary for the `filtered_words`. The key is the word and value is an id that represents the word. 
vocab = {word:idx for idx, word in enumerate(filtered_words, start = 1) }
# Reverse of the `vocab` dictionary. The key is word id and value is the word. 
id2vocab = {idx:word for word, idx in vocab.items()}
# tokenized with the words not in `filtered_words` removed.
filtered = [[token for token in message if token in vocab] for message in tokenized]

### Balancing the classes
Let's do a few last pre-processing steps. If we look at how our twits are labeled, we'll find that 50% of them are neutral. This means that our network will be 50% accurate just by guessing 0 every single time. To help our network learn appropriately, we'll want to balance our classes.
That is, make sure each of our different sentiment scores show up roughly as frequently in the data.

What we can do here is go through each of our examples and randomly drop twits with neutral sentiment. What should be the probability we drop these twits if we want to get around 20% neutral twits starting at 50% neutral? We should also take this opportunity to remove messages with length 0.

In [11]:
balanced = {'messages': [], 'sentiments':[]}

n_neutral = sum(1 for each in sentiments if each == 2)
N_examples = len(sentiments)
keep_prob = (N_examples - n_neutral)/4/n_neutral

for idx, sentiment in enumerate(sentiments):
    message = filtered[idx]
    if len(message) == 0:
        # skip this message because it has length zero
        continue
    elif sentiment != 2 or random.random() < keep_prob:
        balanced['messages'].append(message)
        balanced['sentiments'].append(sentiment) 

If you did it correctly, you should see the following result 

In [12]:
n_neutral = sum(1 for each in balanced['sentiments'] if each == 2)
N_examples = len(balanced['sentiments'])
n_neutral/N_examples

0.19549096401582083

Finally let's convert our tokens into integer ids which we can pass to the network.

In [13]:
token_ids = [[vocab[word] for word in message] for message in balanced['messages']]
sentiments = balanced['sentiments']

In [14]:
token_ids

[[1, 2, 3, 4, 5],
 [6, 7, 8, 9, 10, 11, 12, 13, 14, 3, 15, 16, 17, 18, 2],
 [19, 20, 21, 22, 23, 24, 22, 25, 26, 23, 27, 28],
 [31, 32, 33, 34, 35, 36, 37, 38],
 [39, 40, 41, 42, 43, 44],
 [45, 46, 47, 48, 49, 50, 14, 51, 52, 53],
 [54, 55, 47, 56, 57, 58, 59],
 [32, 60, 61, 62, 63, 64, 65, 66, 67, 58, 64, 68, 69, 70, 71, 72],
 [73, 2],
 [74, 75, 3, 74, 76, 77],
 [14, 78, 79, 80, 81, 82, 83, 84, 85, 86],
 [87, 88, 89, 90, 73, 2, 91, 92, 93, 94, 95, 96],
 [97, 98, 99, 100],
 [101, 40, 41, 42, 43, 44],
 [102, 103, 88, 104, 105, 106, 2, 36, 107, 108, 109, 110, 111, 112, 113],
 [6, 119, 120, 121, 14, 13, 9, 10, 2, 12, 13, 14, 3, 15, 16, 17, 18, 2],
 [25,
  122,
  123,
  124,
  71,
  125,
  126,
  89,
  127,
  9,
  99,
  66,
  128,
  129,
  130,
  20,
  131,
  132,
  133],
 [114, 99, 75, 132, 134, 135, 114, 65, 136, 137, 138, 139, 140, 141, 93, 142],
 [143, 144, 145, 146, 147, 148, 149],
 [150, 74],
 [151, 152, 153, 61, 154, 155],
 [165, 166, 109],
 [64, 112, 167, 153, 61, 168, 169, 170, 17

## Neural Network
Now we have our vocabulary which means we can transform our tokens into ids, which are then passed to our network. So, let's define the network now!

Here is a nice diagram showing the network we'd like to build: 

#### Embed -> RNN -> Dense -> Softmax
### Implement the text classifier
Before we build text classifier, if you remember from the other network that you built in  "Sentiment Analysis with an RNN"  exercise  - which there, the network called " SentimentRNN", here we named it "TextClassifer" - consists of three main parts: 1) init function `__init__` 2) forward pass `forward`  3) hidden state `init_hidden`. 

This network is pretty similar to the network you built expect in the  `forward` pass, we use softmax instead of sigmoid. The reason we are not using sigmoid is that the output of NN is not a binary. In our network, sentiment scores have 5 possible outcomes. We are looking for an outcome with the highest probability thus softmax is a better choice.

In [45]:
# First checking if GPU is available
train_on_gpu = torch.cuda.is_available()

if(train_on_gpu):
    print('[ info ] Training_on_gpu.')
else:
    print('[ info ] No GPU available, training in CPU.')

class TextClassifier(nn.Module):
    def __init__(self, vocab_size, embed_size, lstm_size, output_size, lstm_layers=1, dropout=0.1):
        """
        Initialize the model by setting up the layers.
        
        Parameters
        ----------
            vocab_size : The vocabulary size.
            embed_size : The embedding layer size.
            lstm_size : The LSTM layer size.
            output_size : The output size.
            lstm_layers : The number of LSTM layers.
            dropout : The dropout probability.
        """
        
        super().__init__()
        self.vocab_size = vocab_size
        self.embed_size = embed_size
        self.lstm_size = lstm_size
        self.output_size = output_size
        self.lstm_layers = lstm_layers
        self.dropout = dropout
        
        # TODO Implement

        # Setup embedding layer
        self.embedding = nn.Embedding(vocab_size, embed_size)
        self.lstm = nn.LSTM(embed_size, lstm_size, lstm_layers, 
                            dropout = dropout, batch_first = False)

        
        # Setup additional layers
        # drop out here
        self.drp = nn.Dropout(dropout)
        # linear and and softmax layers
        self.fc = nn.Linear(self.lstm_size, self.output_size)
        self.soft = nn.LogSoftmax()


    def init_hidden(self, batch_size):
        """ 
        Initializes hidden state
        
        Parameters
        ----------
            batch_size : The size of batches.
        
        Returns
        -------
            hidden_state
            
        """
        
        # TODO Implement 
        
        # Create two new tensors with sizes n_layers x batch_size x hidden_dim,
        # initialized to zero, for hidden state and cell state of LSTM
        weight = next(self.parameters()).data
        
        #if (train_on_gpu):
        #    hidden = (weight.new(self.lstm_layers, batch_size, self.lstm_size).zero_().cuda(),
        #              weight.new(self.lstm_layers, batch_size, self.lstm_size).zero_().cuda()) 
        #else:
        
        hidden = (weight.new(self.lstm_layers, batch_size, self.lstm_size).zero_(), 
                      weight.new(self.lstm_layers, batch_size, self.lstm_size).zero_())
            
        return hidden


    def forward(self, nn_input, hidden_state):
        """
        Perform a forward pass of our model on nn_input.
        
        Parameters
        ----------
            nn_input : The batch of input to the NN.
            hidden_state : The LSTM hidden state.

        Returns
        -------
            logps: log softmax output
            hidden_state: The new hidden state.

        """
        
        # TODO Implement
        batch_size = nn_input.size(1)
        
        #embeddings and lstm_out
        nn_input = nn_input.long()
        embeds = self.embedding(nn_input)
        lstm_out, hidden = self.lstm(embeds, hidden_state)
        
        # stack up lstm outputs
        #lstm_out = lstm_out.contiguous().view(-1, self.lstm_size)
        lstm_out = lstm_out[-1,:,:]
        
        # dropout later
        out = self.drp(lstm_out)
        
        # fully connected layer
        out = self.fc(out)
        
        # softmax function
        soft_out = self.soft(out)
        
        # reshape the batch size first
        #soft_out = soft_out.view(batch_size, -1)
        #soft_out = soft_out[:, -1] # get last batch of labels
        
        return soft_out, hidden

[ info ] Training_on_gpu.


In [46]:
# First checking if GPU is available
train_on_gpu = torch.cuda.is_available()

if(train_on_gpu):
    print('[ info ] Training_on_gpu.')
else:
    print('[ info ] No GPU available, training in CPU.')

[ info ] Training_on_gpu.


### View Model

In [47]:
model = TextClassifier(len(vocab), 10, 6, 5, dropout=0.1, lstm_layers=2)
model.embedding.weight.data.uniform_(-1, 1)
input = torch.randint(0, 1000, (5, 4), dtype=torch.int64)
hidden = model.init_hidden(4)
logps, _ = model.forward(input, hidden)
print('[ info ] logps', logps)

[ info ] logps tensor([[-1.7760, -1.3941, -1.6425, -1.7051, -1.5731],
        [-1.7823, -1.4156, -1.6401, -1.6798, -1.5674],
        [-1.7775, -1.3966, -1.6458, -1.7069, -1.5644],
        [-1.7472, -1.4276, -1.6272, -1.6993, -1.5771]])


/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:114: UserWarning: Implicit dimension choice for log_softmax has been deprecated. Change the call to include dim=X as an argument.


## Training
### DataLoaders and Batching
Now we should build a generator that we can use to loop through our data. It'll be more efficient if we can pass our sequences in as batches. Our input tensors should look like `(sequence_length, batch_size)`. So if our sequences are 40 tokens long and we pass in 25 sequences, then we'd have an input size of `(40, 25)`.

If we set our sequence length to 40, what do we do with messages that are more or less than 40 tokens? For messages with fewer than 40 tokens, we will pad the empty spots with zeros. We should be sure to **left** pad so that the RNN starts from nothing before going through the data. If the message has 20 tokens, then the first 20 spots of our 40 long sequence will be 0. If a message has more than 40 tokens, we'll just keep the first 40 tokens.

In [48]:
def dataloader(messages, labels, sequence_length=30, batch_size=32, shuffle=False):
    """ 
    Build a dataloader.
    """
    if shuffle:
        indices = list(range(len(messages)))
        random.shuffle(indices)
        messages = [messages[idx] for idx in indices]
        labels = [labels[idx] for idx in indices]

    total_sequences = len(messages)

    for ii in range(0, total_sequences, batch_size):
        batch_messages = messages[ii: ii+batch_size]
        
        # First initialize a tensor of all zeros
        batch = torch.zeros((sequence_length, len(batch_messages)), dtype=torch.int64)
        for batch_num, tokens in enumerate(batch_messages):
            token_tensor = torch.tensor(tokens)
            # Left pad!
            start_idx = max(sequence_length - len(token_tensor), 0)
            batch[start_idx:, batch_num] = token_tensor[:sequence_length]
        
        label_tensor = torch.tensor(labels[ii: ii+len(batch_messages)])
        
        yield batch, label_tensor

### Training and  Validation
With our data in nice shape, we'll split it into training and validation sets.

In [49]:
"""
Split data into training and validation datasets. Use an appropriate split size.
The features are the `token_ids` and the labels are the `sentiments`.
"""   

split_perc = 0.75

# TODO Implement 
split_idx = int(len(token_ids)*split_perc)

train_features = token_ids[:split_idx]
valid_features = token_ids[split_idx:]
train_labels = sentiments[:split_idx]
valid_labels = sentiments[split_idx:]

In [50]:
text_batch, labels = next(iter(dataloader(train_features, train_labels, sequence_length=20, batch_size=64)))
model = TextClassifier(len(vocab)+1, 200, 128, 5, dropout=0.)
hidden = model.init_hidden(64)
logps, hidden = model.forward(text_batch, hidden)

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:114: UserWarning: Implicit dimension choice for log_softmax has been deprecated. Change the call to include dim=X as an argument.


### Training
It's time to train the neural network!

In [55]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

model = TextClassifier(len(vocab)+1, 1024, 512, 5, lstm_layers=2, dropout=0.2)
model.embedding.weight.data.uniform_(-1, 1)
model.to(device)

TextClassifier(
  (embedding): Embedding(99355, 1024)
  (lstm): LSTM(1024, 512, num_layers=2, dropout=0.2)
  (drp): Dropout(p=0.2)
  (fc): Linear(in_features=512, out_features=5, bias=True)
  (soft): LogSoftmax()
)

In [56]:
"""
Train your model with dropout. Make sure to clip your gradients.
Print the training loss, validation loss, and validation accuracy for every 100 steps.
"""

epochs = 3
batch_size = 512
learning_rate = 0.001

print_every = 100
criterion = nn.NLLLoss() # 
optimizer = optim.Adam(model.parameters(), lr=learning_rate)

# train
model.train()

for epoch in range(epochs):
    print('Starting epoch {}'.format(epoch + 1))
    steps = 0
    for text_batch, labels in dataloader(
            train_features, train_labels, batch_size=batch_size):
        steps += 1
        hidden = model.init_hidden(labels.shape[0])
        
        # Set Device
        text_batch, labels = text_batch.to(device), labels.to(device)
        
        for each in hidden:
            each.to(device)
        
        # TODO Implement: Train Model
        # zero accumulated gradients
        model.zero_grad()
        
        # Get the output from the model
        output, hidden = model(text_batch, hidden)
        
        # calculate the loss and perform backprop
        loss = criterion(output.squeeze(), labels)
        loss.backward()
        
        # clip_grad_norm helps prevent the exploding gradient problem in RNNs
        nn.utils.clip_grad_norm_(model.parameters(), 5)
        optimizer.step()
        
        if steps % print_every == 0:
            # Get validation loss
            model.eval()
            val_losses = [] # Keep track of the losses for iteration...
            for inputs, labels in dataloader(valid_features, valid_labels, batch_size = batch_size):
                # move the inputs and labels to the relevant device
                inputs, labels = inputs.to(device), labels.to(device)
                
                val_h = model.init_hidden(labels.shape[0])
                
                for each in val_h:
                    each.to(device)
                
                output, val_h = model(inputs, val_h)
                val_loss = criterion(output.squeeze(), labels)
                val_losses.append(val_loss.item())
                
            # TODO Implement: Print metrics
            model.train()
            print('[ info ] Epoch: {}/{}...'.format(epoch + 1, epochs),
                 '[ info ] Step: {}...'.format(steps),
                 '[ info ] Loss: {:.6f}...'.format(loss.item()),
                 '[ info ] Validation Loss: {:.6f}'.format(np.mean(val_losses)))
            

Starting epoch 1


/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:114: UserWarning: Implicit dimension choice for log_softmax has been deprecated. Change the call to include dim=X as an argument.


[ info ] Epoch: 1/3... [ info ] Step: 100... [ info ] Loss: 1.066079... [ info ] Validation Loss: 1.072607
[ info ] Epoch: 1/3... [ info ] Step: 200... [ info ] Loss: 0.674735... [ info ] Validation Loss: 0.972863
[ info ] Epoch: 1/3... [ info ] Step: 300... [ info ] Loss: 0.844146... [ info ] Validation Loss: 0.894779
[ info ] Epoch: 1/3... [ info ] Step: 400... [ info ] Loss: 0.861928... [ info ] Validation Loss: 0.861061
[ info ] Epoch: 1/3... [ info ] Step: 500... [ info ] Loss: 0.829485... [ info ] Validation Loss: 0.825909
[ info ] Epoch: 1/3... [ info ] Step: 600... [ info ] Loss: 0.537716... [ info ] Validation Loss: 0.805902
[ info ] Epoch: 1/3... [ info ] Step: 700... [ info ] Loss: 0.918092... [ info ] Validation Loss: 0.786116
[ info ] Epoch: 1/3... [ info ] Step: 800... [ info ] Loss: 0.725815... [ info ] Validation Loss: 0.782783
[ info ] Epoch: 1/3... [ info ] Step: 900... [ info ] Loss: 0.823131... [ info ] Validation Loss: 0.783761
[ info ] Epoch: 1/3... [ info ] Step:

## Making Predictions
### Prediction 
Okay, now that you have a trained model, try it on some new twits and see if it works appropriately. Remember that for any new text, you'll need to preprocess it first before passing it to the network. Implement the `predict` function to generate the prediction vector from a message.

In [65]:
def predict(text, model, vocab):
    """ 
    Make a prediction on a single sentence.

    Parameters
    ----------
        text : The string to make a prediction on.
        model : The model to use for making the prediction.
        vocab : Dictionary for word to word ids. The key is the word and the value is the word id.

    Returns
    -------
        pred : Prediction vector
    """    
    
    # TODO Implement
    
    tokens = preprocess(text)
    
    # Filter non-vocab words, just like above
    tokens = [token for token in tokens if token in vocab]
    
    # Convert words to ids
    tokens = [vocab[token] for token in tokens]
        
    # Adding a batch dimension
    text_input = torch.FloatTensor(tokens).view(-1,1)
    text_input = torch.from_numpy(np.asarray(text_input))
    
    # Get the NN output
    hidden = model.init_hidden(1)
    logps, _ = model.forward(text_input, hidden)
    # Take the exponent of the NN output to get a range of 0 to 1 for each label.
    pred = torch.exp(logps)
    
    return pred

In [66]:
text = "Google is working on self driving cars, I'm bullish on $goog"
model.eval()
model.to("cpu")
predict(text, model, vocab)

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:114: UserWarning: Implicit dimension choice for log_softmax has been deprecated. Change the call to include dim=X as an argument.


tensor([[ 0.0001,  0.0198,  0.0040,  0.8509,  0.1252]])

### Questions: What is the prediction of the model? What is the uncertainty of the prediction?
** TODO: Answer Question**

"""The model is predicting a high probability (85%) of the text being positive and being a sentiment of 1.
The uncertainly of this prediction being a 1 is 15% which is quite low. I also think if we combine the probabilities
on the positive side (85%+12.5%) the model predicts a positive sentiment overall. Which we agree with
intuitively based on the text."""

Now we have a trained model and we can make predictions. We can use this model to track the sentiments of various stocks by predicting the sentiments of twits as they are coming in. Now we have a stream of twits. For each of those twits, pull out the stocks mentioned in them and keep track of the sentiments. Remember that in the twits, ticker symbols are encoded with a dollar sign as the first character, all caps, and 2-4 letters, like $AAPL. Ideally, you'd want to track the sentiments of the stocks in your universe and use this as a signal in your larger model(s).

## Testing
### Load the Data 

In [70]:
with open(os.path.join('..', '..', 'data', 'project_6_stocktwits', 'test_twits.json'), 'r') as f:
    test_data = json.load(f)

In [71]:
test_data

{'data': [{'message_body': '$JWN has moved -1.69% on 10-31. Check out the movement and peers at  https://dividendbot.com?s=JWN',
   'timestamp': '2018-11-01T00:00:05Z'},
  {'message_body': '$MCD bt puts when Cramer touted going to $200 so for a profit and bt back yesterday down on up day is good $PEP looks lower too',
   'timestamp': '2018-11-01T00:00:05Z'},
  {'message_body': '$AAPL 190 after ER', 'timestamp': '2018-11-01T00:00:18Z'},
  {'message_body': '$PYPL Insider &quot;Pentland Adele Louise&quot; just cashed in 8,149 Options (Tran Code - M, S)  --- Insider Forms, iOS app',
   'timestamp': '2018-11-01T00:00:25Z'},
  {'message_body': '$AMZN A lot of technical signs point to a short-term bottom in place for stocks (John Murphy)',
   'timestamp': '2018-11-01T00:00:42Z'},
  {'message_body': '$AAPL https://stockcharts.com/freecharts/pnf.php?c=aapl,PWTAWANRBO[PA][D][F1!3!!!2!20]',
   'timestamp': '2018-11-01T00:00:55Z'},
  {'message_body': '$NFLX Option Order Flow Sentiment is 60.9% Bul

### Twit Stream

In [72]:
def twit_stream():
    for twit in test_data['data']:
        yield twit

next(twit_stream())

{'message_body': '$JWN has moved -1.69% on 10-31. Check out the movement and peers at  https://dividendbot.com?s=JWN',
 'timestamp': '2018-11-01T00:00:05Z'}

Using the `prediction` function, let's apply it to a stream of twits.

In [73]:
def score_twits(stream, model, vocab, universe):
    """ 
    Given a stream of twits and a universe of tickers, return sentiment scores for tickers in the universe.
    """
    for twit in stream:

        # Get the message text
        text = twit['message_body']
        symbols = re.findall('\$[A-Z]{2,4}', text)
        score = predict(text, model, vocab)

        for symbol in symbols:
            if symbol in universe:
                yield {'symbol': symbol, 'score': score, 'timestamp': twit['timestamp']}

In [74]:
universe = {'$BBRY', '$AAPL', '$AMZN', '$BABA', '$YHOO', '$LQMT', '$FB', '$GOOG', '$BBBY', '$JNUG', '$SBUX', '$MU'}
score_stream = score_twits(twit_stream(), model, vocab, universe)

next(score_stream)

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:114: UserWarning: Implicit dimension choice for log_softmax has been deprecated. Change the call to include dim=X as an argument.


{'symbol': '$AAPL',
 'score': tensor([[ 0.1758,  0.2068,  0.1394,  0.1632,  0.3148]]),
 'timestamp': '2018-11-01T00:00:18Z'}

That's it. You have successfully built a model for sentiment analysis! 

## Submission
Now that you're done with the project, it's time to submit it. Click the submit button in the bottom right. One of our reviewers will give you feedback on your project with a pass or not passed grade. You can continue to the next section while you wait for feedback.